### ISSUE TITLE
nginx: [emerg] "modsecurity_rules_file" directive Rules error. File: /usr/local/owasp-modsecurity-crs/rules/REQUEST-914-FILE-DETECTION.conf. Line: 71. Column: 85. 

### ISSUE THREAD

**Issue by ghost** [SpiderLabs/owasp-modsecurity-crs#1217]
nginx: [emerg] "modsecurity_rules_file" directive Rules error. File: /usr/local/owasp-modsecurity-crs/rules/REQUEST-914-FILE-DETECTION.conf. Line: 71. Column: 85. Expecting a variable, got:  :  S}: %{MATCHED_VARS}',\ in /etc/nginx/nginx.conf:28
nginx: configuration file /etc/nginx/nginx.conf test failed

Nginx Version - 1.15.5
Mod-security - v3/master
Owasp Rules - 3.2/dev

**Comment by zimmerle** [comment]
Hi @Shubham-Panwar,

This is actually a problem in ModSecurity. The parser needs to understand two variables in a sequence, which is not yet the case. In the meanwhile, you can put any character in between the two variables in question.

**Comment by ghost** [comment]
can you please tell me exactly where do i need to put the character here  ? 
  
```
logdata:'Matched Data: %{TX.0} found within %{MATCHED_VAR_NAMES}: %{MATCHED_VARS}',\
    t:none,\
    tag:'application-multi',\
    tag:'platform-multi',\
    tag:'attack-file-upload',\
    tag:'OWASP_CRS/WEB_ATTACK/MALICIOUS_FILE',\
    tag:'paranoia-level/1',\
    ver:'OWASP_CRS/3.1.0',\
    severity:'NOTICE',\
    setvar:'tx.msg=%{rule.msg}',\
    setvar:'tx.anomaly_score=+%{tx.notice_anomaly_score}',\
    setvar:'tx.%{rule.id}-OWASP_CRS/WEB_ATTACK/MALICIOUS-FILE-%{matched_var_name}=%{tx.0}'
```

**Comment by zimmerle** [comment]
I messed up everything. My fault.

The problem here is that there is no ***MATCHED_VAR_NAMES***  in ModSecurity. The correct name is ***MATCHED_VAR_NAME*** or ***MATCHED_VARS_NAMES***. 


**Implementation on version 2**:

https://github.com/SpiderLabs/ModSecurity/blob/49495f1925a14f74f93cb0ef01172e5abc3e4c55/apache2/re_variables.c#L912-L922

https://github.com/SpiderLabs/ModSecurity/blob/49495f1925a14f74f93cb0ef01172e5abc3e4c55/apache2/re_variables.c#L2129-L2192


**Implementation on version 3**:

https://github.com/SpiderLabs/ModSecurity/blob/18cdffdbca75e6b9f790f6df2807a32cd805c0a0/src/variables/matched_var_name.h#L33

https://github.com/SpiderLabs/ModSecurity/blob/18cdffdbca75e6b9f790f6df2807a32cd805c0a0/src/variables/matched_vars_names.h#L33-L34

The link on the manual:
- https://github.com/SpiderLabs/ModSecurity/wiki/Reference-Manual-%28v2.x%29#matched_var_name
- https://github.com/SpiderLabs/ModSecurity/wiki/Reference-Manual-%28v2.x%29#matched_vars_names

In others words, this is not workable in any ModSecurity version. The difference is that on version 2 it will fail silent. The problem is easy to catch if you test this rule, the _msg_ will not be generated as expected. Thanks to version 3 we are able to spot this not so well written rules up front, without leading to a problem on run time.

Use this instead (not tested):
```
logdata:'Matched Data: %{TX.0} found within %{MATCHED_VARS_NAMES}: %{MATCHED_VARS}',\
    t:none,\
    tag:'application-multi',\
    tag:'platform-multi',\
    tag:'attack-file-upload',\
    tag:'OWASP_CRS/WEB_ATTACK/MALICIOUS_FILE',\
    tag:'paranoia-level/1',\
    ver:'OWASP_CRS/3.1.0',\
    severity:'NOTICE',\
    setvar:'tx.msg=%{rule.msg}',\
    setvar:'tx.anomaly_score=+%{tx.notice_anomaly_score}',\
    setvar:'tx.%{rule.id}-OWASP_CRS/WEB_ATTACK/MALICIOUS-FILE-%{matched_var_name}=%{tx.0}'
```

I would strongly recommend to the community of this project to improve their QA by adopting an continuous integration. ModSecurity version 3 already provide all the tools such as the [rules-check](https://github.com/SpiderLabs/ModSecurity/blob/v3/master/tools/rules-check/rules-check.cc) utility.

**Comment by fzipi** [comment]
@zimmerle We kind of have that already in https://github.com/CRS-support/secrules_parsing. And is v2 and v3 compatible.

**Comment by fzipi** [comment]
This is fixed in #1214 .

**Comment by fzipi** [comment]
Applied fix.

**Comment by zimmerle** [comment]
> @zimmerle We kind of have that already in https://github.com/CRS-support/secrules_parsing. And is v2 and v3 compatible.

Good to know. In addition, you may want to also consider to use the ModSecurity rules-check utility. As it seems to catch that cases that is not being catch by the that tool. It servers well both: v2 and v3 as they are compatible rule-wise.

**Comment by victorhora** [comment]
@fzipi 

I think #1214 could be a duplicate of what's already included in PR https://github.com/SpiderLabs/owasp-modsecurity-crs/pull/1187. That PR should also fix typos in other places.

**Comment by fzipi** [comment]
@victorhora I see. Do you think I should revert then?

**Comment by victorhora** [comment]
Hummm, I'm no expert in Git, but maybe the easier thing to do might be leaving #1214 as is and then just change #1187 accordingly before merging. I think GitHub should be able to handle this as well.

My comment was more of a reminder that we may still have issues with these things and being so consider merging #1187 in addition to #1214 @fzipi ;)

**Comment by fzipi** [comment]
So this was fixed in #1229 , and backported to v3.1/dev in #1233.
I'm closing this one now, and will create a new issue to consider using the ModSecurity rules-check utility.
